File: photoDemo/photoDemo/ViewController.m
//
//  ViewController.m
//  photoDemo
//
//  Created by zm004 on 16/4/18.
//  Copyright © 2016年 zm004. All rights reserved.
//

#import "ViewController.h"
@import AssetsLibrary;
@interface ViewController ()
{

    UIImageView *imageView;
}
@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    _images = [[NSMutableArray alloc] init];
    [self reloadImagesFromLibrary];
    
    UIButton *button = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [button setFrame:CGRectMake(50, 50, 100, 30)];
    [button setTitle:@"Get Images" forState:UIControlStateNormal];
    [button addTarget:self action:@selector(logImages) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:button];
    
    imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 100, self.view.frame.size.width, self.view.frame.size.height - 100)];
    [imageView setContentMode:UIViewContentModeScaleAspectFit];
    [self.view addSubview:imageView];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)logImages
{
    _images = (NSMutableArray*)[self images];
    NSLog(@"images is %@",_images);
    [self getImage:[_images objectAtIndex:10]];
}

- (void)getImage:(NSString *)urlStr
{
    ALAssetsLibrary *assetLibrary=[[ALAssetsLibrary alloc] init];
    NSURL *url=[NSURL URLWithString:urlStr];
    [assetLibrary assetForURL:url resultBlock:^(ALAsset *asset)  {
        UIImage *image=[UIImage imageWithCGImage:asset.thumbnail];
        imageView.image=image;
        
    }failureBlock:^(NSError *error) {
        NSLog(@"error=%@",error);
    }];
}


- (void)reloadImagesFromLibrary
{
    self.images = [[NSMutableArray alloc] init];
    dispatch_async(dispatch_get_main_queue(), ^{
        
        @autoreleasepool {
            ALAssetsLibraryAccessFailureBlock failureblock = ^(NSError *myerror){
                NSLog(@"相册访问失败 =%@", [myerror localizedDescription]);
                if ([myerror.localizedDescription rangeOfString:@"Global denied access"].location!=NSNotFound) {
                    NSLog(@"无法访问相册.请在'设置->定位服务'设置为打开状态.");
                }else{
                    NSLog(@"相册访问失败.");
                }
            };
            
            ALAssetsGroupEnumerationResultsBlock groupEnumerAtion = ^(ALAsset *result, NSUInteger index, BOOL *stop){
                if (result!=NULL) {
                    
                    if ([[result valueForProperty:ALAssetPropertyType] isEqualToString:ALAssetTypePhoto]) {
                        
                        NSString *urlstr=[NSString stringWithFormat:@"%@",result.defaultRepresentation.url];//图片的url
                        [self.images addObject:urlstr];
                        //NSLog(@"urlStr is %@",urlstr);
                        /*result.defaultRepresentation.fullScreenImage//图片的大图
                         result.thumbnail                             //图片的缩略图小图
                         //                    NSRange range1=[urlstr rangeOfString:@"id="];
                         //                    NSString *resultName=[urlstr substringFromIndex:range1.location+3];
                         //                    resultName=[resultName stringByReplacingOccurrencesOfString:@"&ext=" withString:@"."];//格式demo:123456.png
                         */
                    }
                }
            };
            
            ALAssetsLibraryGroupsEnumerationResultsBlock libraryGroupsEnumeration = ^(ALAssetsGroup* group, BOOL* stop){
                
                if (group == nil)
                {
                    
                }
                
                if (group!=nil) {
                    NSString *g=[NSString stringWithFormat:@"%@",group];//获取相簿的组
                    NSLog(@"gg:%@",g);//gg:ALAssetsGroup - Name:Camera Roll, Type:Saved Photos, Assets count:71
                    
                    NSString *g1=[g substringFromIndex:16 ] ;
                    NSArray *arr=[[NSArray alloc] init];
                    arr=[g1 componentsSeparatedByString:@","];
                    NSString *g2=[[arr objectAtIndex:0] substringFromIndex:5];
                    if ([g2 isEqualToString:@"Camera Roll"]) {
                        g2=@"相机胶卷";
                    }
                    NSString *groupName=g2;//组的name
                    
                    [group enumerateAssetsUsingBlock:groupEnumerAtion];
                }
                
            };
            
            ALAssetsLibrary* library = [[ALAssetsLibrary alloc] init];
            [library enumerateGroupsWithTypes:ALAssetsGroupAll
                                   usingBlock:libraryGroupsEnumeration
                                 failureBlock:failureblock];
        }
        
    });
}
@end

File: storeManageTest/storeManageTest/ViewController.swift
//
//  ViewController.swift
//  storeManageTest
//
//  Created by zm004 on 16/7/12.
//  Copyright © 2016年 zm004. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    var a : A?
    var c : C?
    var b : B?
    override func viewDidLoad() {
        super.viewDidLoad()
        a = A()
//        c = C()
        
        b = B()
        a = nil
        
//        a?.printName()
//        a?.unownedPrint()
        
//        c?.sameClosToA = a!.printName
//        a = nil
//        c!.sameClosToA()
        
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}
class A :NSObject, RequestHandler {
    var b : B!
    lazy var printName:()->() = {
        [weak self] in
        if let strongSelf = self{
            print("闭包: \(self!.b)")
        }
    }
    lazy var unownedPrint:()->() = {
        [unowned self] in
        print("test")
        print("unowned闭包: \(self.b)")
    }
    override init(){
        super.init()
        b = B()
        b.delegate = self
    }
    func requestFinished() {
        print("协议中的方法")
    }
    deinit{
        print("A deinit")
    }
}
protocol RequestHandler : NSObjectProtocol{
    func requestFinished()
}
class B {
    weak var delegate : RequestHandler!
    deinit{
        print("B deinit")
    }
}
class C {
    var sameClosToA : (()->())!
    init(){}
}